### PR TITLE
Update region computation.

### PR DESCRIPTION
**Pull request by nimakarimipour**
#### CI
~~Please note that CI will fail for this PR as it requires release of `NullAway:0.10.5`.~~
`NullAway:0.10.5` is released and CI should pass.

-----

This PR updates the region computation in `Type-Annotator-Scanner` according to latest changes in NullAway following [PR658](https://github.com/uber/NullAway/pull/658). 

Also in this PR the infrastructure for keeping compatibility with older versions of NullAway (`0.10.4` or less) is implemented.

### CODE REVIEW

**Review by msridhar — COMMENTED** [review]
What is here LGTM, though I know we plan to update this one to keep compatibility with NullAway 0.10.4.  I guess we can approve after this PR is finalized.

File: core/src/main/java/edu/ucr/cs/riple/core/adapters/Adapter.java
Comment (msridhar) [inline core/src/main/java/edu/ucr/cs/riple/core/adapters/Adapter.java:39]: Let's be more verbose here and call this `NullAwayVersionAdapter`

File: core/src/main/java/edu/ucr/cs/riple/core/adapters/AdapterAbstractClass.java
Comment (msridhar) [inline core/src/main/java/edu/ucr/cs/riple/core/adapters/AdapterAbstractClass.java:30]: I don't think it's worth creating this abstract class just to hold a single protected field.  I would get rid of it.

File: core/src/main/java/edu/ucr/cs/riple/core/adapters/NullAwayAdapterVersion0.java
Comment (msridhar) [inline core/src/main/java/edu/ucr/cs/riple/core/adapters/NullAwayAdapterVersion0.java:48]: If we go with `NullAwayVersionAdapter` for the interface, this can be `NullAwayVersion0Adapter`, or simply `NullAwayV0Adapter`

File: core/src/main/java/edu/ucr/cs/riple/core/adapters/NullAwayAdapterVersion1.java
Comment (msridhar) [inline core/src/main/java/edu/ucr/cs/riple/core/adapters/NullAwayAdapterVersion1.java:48]: Similar comment about class naming

File: core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/Region.java
Comment (msridhar) [inline core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/Region.java:44]:
```suggestion
  /** Different types of code segments for a region. */
```

**Review by msridhar — COMMENTED** [review]
Mostly LGTM, just a few comments

File: core/src/main/java/edu/ucr/cs/riple/core/adapters/Adapter.java
Comment (nimakarimipour) [reply core/src/main/java/edu/ucr/cs/riple/core/adapters/Adapter.java:39]: aa68932d9024f74a03572fd0ac452a4e7841ca2f

File: core/src/main/java/edu/ucr/cs/riple/core/adapters/AdapterAbstractClass.java
Comment (nimakarimipour) [reply core/src/main/java/edu/ucr/cs/riple/core/adapters/AdapterAbstractClass.java:30]: Deleted in: aa68932d9024f74a03572fd0ac452a4e7841ca2f

File: core/src/main/java/edu/ucr/cs/riple/core/adapters/NullAwayAdapterVersion0.java
Comment (nimakarimipour) [reply core/src/main/java/edu/ucr/cs/riple/core/adapters/NullAwayAdapterVersion0.java:48]: Sure, sound good. aa68932d9024f74a03572fd0ac452a4e7841ca2f

Thank you for the suggestions in renaming.

File: core/src/main/java/edu/ucr/cs/riple/core/adapters/NullAwayAdapterVersion1.java
Comment (nimakarimipour) [reply core/src/main/java/edu/ucr/cs/riple/core/adapters/NullAwayAdapterVersion1.java]: b51475a6266c9f9db8b9ccc48dd839e8dc121cd9

**Comment by nimakarimipour** [comment]
@msridhar Thank you for the review. I resolved all comments.

File: core/src/main/java/edu/ucr/cs/riple/core/metadata/field/FieldInitializationAnalysis.java
Comment (lazaroclapp) [inline core/src/main/java/edu/ucr/cs/riple/core/metadata/field/FieldInitializationAnalysis.java:30]: Could we extract `"field_init.tsv"` into a global constant, like we do below for `MethodDeclarationTree`:

```
    super(config, config.target.dir.resolve(Serializer.METHOD_INFO_FILE_NAME));
```

**Review by lazaroclapp — COMMENTED** [review]
Would like to see CI passing on the recently released NullAway 0.10.5 before landing this, but here are my comments so far :)

File: .github/workflows/continuous-integration.yml
Comment (nimakarimipour) [reply .github/workflows/continuous-integration.yml:59]: #98

File: .github/workflows/continuous-integration.yml
Comment (nimakarimipour) [reply .github/workflows/continuous-integration.yml]: Sure makes sense. 69e537b6295ef1f21e3e7b7f6abcd28a2b5728c1

File: type-annotator-scanner/src/test/java/edu/ucr/cs/riple/scanner/tools/TrackerNodeDisplay.java
Comment (nimakarimipour) [reply type-annotator-scanner/src/test/java/edu/ucr/cs/riple/scanner/tools/TrackerNodeDisplay.java]: Yes, they it should be `calleeMember` and yes it can be a field as well. I understand that `calleeMember` is not accurate as it can be a `field` as well. I am going to change it for now to `calleeMember` but it should be changed in future. 851f8ebc67f28bf42122f9a51a4df1f57639545c

Comment (msridhar) [reply type-annotator-scanner/src/test/java/edu/ucr/cs/riple/scanner/tools/TrackerNodeDisplay.java]: I _think_ "caller" here refers to the region containing a use of some member, and "callee" refers to the used member.  So maybe `regionClass`, `regionMember`, `usedClass`, `usedMember`?  @nimakarimipour if you agree we can just make the change.  Otherwise, we can leave a TODO comment and fix in a follow-up.

File: type-annotator-scanner/src/test/java/edu/ucr/cs/riple/scanner/FieldTrackerTest.java
Comment (nimakarimipour) [reply type-annotator-scanner/src/test/java/edu/ucr/cs/riple/scanner/FieldTrackerTest.java:124]: Methods `matchIdentifier` and `matchMemberSelect` might produce the same output, I just thought it might not worth it to keep all the outputs in a set and check for every output if it is already serialized as they will be deserialized and converted into a set in `core`. Also I didn't want to change the test infra in this PR. Please let me know if you think I should change the infra or make a list and prevent duplicates serialize in output.

File: type-annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/out/TrackerNode.java
Comment (nimakarimipour) [reply type-annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/out/TrackerNode.java:73]: Yes, we are creating `A.b` and `A.a`. In previous approach the enclosing class of these fields (`a` and `b`) was still considered to be `A` and we would have generated region `A.null`. In the new approach we are making `A.a` and `A.b`. I am not sure if they are actually wrong as long as NullAway considers the identical region for triggered error and fix. Please let me know if we should still change this.

File: core/src/main/java/edu/ucr/cs/riple/core/adapters/NullAwayVersionAdapter.java
Comment (nimakarimipour) [reply core/src/main/java/edu/ucr/cs/riple/core/adapters/NullAwayVersionAdapter.java]: 2ae2b00eed9645d44e0890316a98d1ddda83953d

File: core/src/main/java/edu/ucr/cs/riple/core/adapters/NullAwayVersionAdapter.java
Comment (nimakarimipour) [reply core/src/main/java/edu/ucr/cs/riple/core/adapters/NullAwayVersionAdapter.java]: Thank you for noticing this. 204ec6b42decdb845e0a72447bb33d47df21d016

File: core/src/main/java/edu/ucr/cs/riple/core/adapters/NullAwayV0Adapter.java
Comment (nimakarimipour) [reply core/src/main/java/edu/ucr/cs/riple/core/adapters/NullAwayV0Adapter.java:58]: Sure, sounds good. b21407efd6fc3b03f270fe52fa240122658b0809

File: core/src/main/java/edu/ucr/cs/riple/core/metadata/field/FieldInitializationAnalysis.java
Comment (nimakarimipour) [reply core/src/main/java/edu/ucr/cs/riple/core/metadata/field/FieldInitializationAnalysis.java:30]: Actually `"field_init.tsv"` is not an output of `TypeAnnotatorScanner` and it might not be preferred to declare the file name in its `Serializer`. But we should definitely declare the file as a global constant. I extracted it into the user class. caff1d3c011c485fdb178886e55465f29297c22e

File: type-annotator-scanner/src/test/java/edu/ucr/cs/riple/scanner/tools/TrackerNodeDisplay.java
Comment (nimakarimipour) [reply type-annotator-scanner/src/test/java/edu/ucr/cs/riple/scanner/tools/TrackerNodeDisplay.java]: @msridhar Sure, we are already renaming a member of this class in this PR, I will update it now. Thank you for the suggestions.

Comment (nimakarimipour) [reply type-annotator-scanner/src/test/java/edu/ucr/cs/riple/scanner/tools/TrackerNodeDisplay.java]: bc2ad4e1c4dfe927a6a172e03c82140f2908700b

**Comment by nimakarimipour** [comment]
@lazaroclapp Thank you very much for the review, this is ready for another round.

File: type-annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/out/TrackerNode.java
Comment (msridhar) [reply type-annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/out/TrackerNode.java:73]: @nimakarimipour the issue here is that there are no fields `A.a` and `A.b`; `a` and `b` are local variables inside the static initializer.  Ideally, we would not create regions for entities that are not fields.  I would think you could easily add a check that the symbol is actually for a field?  You can check `symbol.getKind().equals(ElementKind.FIELD)`

File: type-annotator-scanner/src/test/java/edu/ucr/cs/riple/scanner/FieldTrackerTest.java
Comment (lazaroclapp) [reply type-annotator-scanner/src/test/java/edu/ucr/cs/riple/scanner/FieldTrackerTest.java:124]: Makes sense to me, just wanted to understand this. I do think eventually you might want tests to not need this repetition, but it's low priority and not needed for this PR.

**Review by lazaroclapp — COMMENTED** [review]
All changes since my last review look good, and most comments addressed, however, [this one](https://github.com/ucr-riple/NullAwayAnnotator/pull/91#discussion_r1026799094) still stands, see @msridhar's note.

File: type-annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/out/TrackerNode.java
Comment (nimakarimipour) [reply type-annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/out/TrackerNode.java:73]: @msridhar @lazaroclapp Sure, it is wrong to create a region for `A.a` and `A.b` conceptually. To address this issue the suggested change will work and it will create correct results. But if we apply this change, it will require a follow up change in NullAway as well. If we do not apply this change, we will not get inaccurate results as `NullAway` and `Type-Annotator-Scanner` uses the same logic to label regions. And if we apply this change here in this PR, we might get inaccurate results using `NullAway:0.10.5`. Please let me know if I should apply the change anyway and make a follow up on NullAway as well. Thank you.

Comment (msridhar) [reply type-annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/out/TrackerNode.java:73]: Agreed with @lazaroclapp; I'd like to understand how the current version will behave if we don't do a fix here.  If the current behavior won't lead to broken generated code or a new kind of compilation error that doesn't get suppressed, then I am ok with tracking and fixing in a future release.

Comment (nimakarimipour) [reply type-annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/out/TrackerNode.java:73]: @msridhar @lazaroclapp It is still capable of examining `@Nullable` annotations on `a` and `b`, however NullAway will not suggest a fix for these variables since they are local variables. 

The only case that will not be able to handle is seeing a dereference error in their initialization expression.
```java
Class Bar {
   static {
        @Nullable Foo a;
        Foo b = a.deref();
    }
}
```
With current labeling, the region for dereferencing `a.bar()` is `Bar#b` and we cannot add a `SuppressWarnings` on that.

But in the suggested version which is labeling this region as `Bar#null` it is not still possible to silence this error with `@SuppressWarnings` as well. (For both cases the only possible solution is to add the `SuppressWarnings` at the class level).

In summary, the current implementation does not produce inaccurate results and it will not leave an error which the proposed approach can resolve.

Comment (lazaroclapp) [reply type-annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/out/TrackerNode.java:73]: > (For both cases the only possible solution is to add the SuppressWarnings at the class level).

So, just to be clear: What is the tool's _current_ action on the code below:

```
public class A {
   private static Object f;
   static {
     Object a = B.foo();
     Object b = B.bar();
     if (Objects.hashCode(a) > Objects.hashCode(b)) {
       f = a;
     } else {
       f = b;
     }
   }
}
public class B {
   @Nullable
   public static Object foo() {...}
   @Nullable
   public static Object bar() {...}
}
```

Changed to `Objects.hashCode(...)` so there isn't a third-party call taking `null` in there...

Comment (nimakarimipour) [reply type-annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/out/TrackerNode.java:73]: @lazaroclapp I ran the example you provided, please find the output below:

```java
public class A {
    @Nullable
    private static Object f;
    static {
        Object a = B.foo();
        Object b = B.bar();
        if (Objects.hashCode(a) > Objects.hashCode(b)) {
            f = a;
        } else {
            f = b;
        }
    }
}
class B {
    @Nullable
    public static Object foo() {return null;}
    @Nullable
    public static Object bar() {return null;}
}
```

Leaving no remaining NullAway error.

Comment (nimakarimipour) [reply type-annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/out/TrackerNode.java:73]: Added a unit test 2218ea39faaba39e5cd79702f40811f9069084df

Comment (msridhar) [reply type-annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/out/TrackerNode.java:73]: > In summary, the current implementation does not produce inaccurate results and it will not leave an error which the proposed approach can resolve.

Ok, in this case, I am fine with going with this change, though we should track fixing this issue of invalid field regions everywhere (NullAway, scanner, annotator) as a follow-up.

**Review by msridhar — APPROVED** [review]

File: core/src/test/java/edu/ucr/cs/riple/core/CoreTest.java
Comment (lazaroclapp) [inline core/src/test/java/edu/ucr/cs/riple/core/CoreTest.java:329]: 👍

**Review by lazaroclapp — APPROVED** [review]
Ok, given the discussion above and the results in the unit test, I think the chance of a case that will break due to the spurious local regions is sufficiently small. This LGTM! Thanks, Nima, and apologies for all the back and forth! 🙂

**Comment by nimakarimipour** [comment]
@msridhar @lazaroclapp Thank you very much for the review. I will land it now.

**Pull request merged**
